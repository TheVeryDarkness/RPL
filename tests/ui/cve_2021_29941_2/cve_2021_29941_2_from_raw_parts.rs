//@compile-flags: -Z inline-mir=false

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
pub fn swap_index(bla: impl ExactSizeIterator<Item = u32>) -> Vec<u32> {
    let len = bla.len();
    let mut vec = Vec::with_capacity(len);
    let arr: &mut [u32] = unsafe { std::slice::from_raw_parts_mut(vec.as_mut_ptr(), len) };
    //~^ERROR: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
    for (i, a) in bla.enumerate() {
        arr[a as usize] = i as u32;
    }

    unsafe {
        vec.set_len(len);
    }
    vec
}

fn main() {}