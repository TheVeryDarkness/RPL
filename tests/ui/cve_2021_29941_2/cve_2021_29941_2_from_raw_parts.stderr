error: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
  --> tests/ui/cve_2021_29941_2/cve_2021_29941_2_from_raw_parts.rs:7:36
   |
LL |     let len = bla.len();
   |               --------- slice created with this length
LL |     let mut vec = Vec::with_capacity(len);
   |                   ----------------------- `std::vec::Vec` created but not initialized
LL |     let arr: &mut [u32] = unsafe { std::slice::from_raw_parts_mut(vec.as_mut_ptr(), len) };
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^----------------^^^^^^
   |                                                                   |
   |                                                                   slice created with this pointer
   |
   = help: See https://doc.rust-lang.org/std/slice/fn.std::slice::from_raw_parts_mut.html

error: aborting due to 1 previous error

